gm8emulator/src/imgui_utils.rs

```diff
@@ -136,77 +215,104 @@ impl UiCustomFunction for imgui::Ui {
-                    0,
-                    Vec2(0.5, 0.5).into(),
-                );
-                c::igBegin("Information\0".as_ptr() as _, std::ptr::null_mut(), 0b0001_0111_1110);
-                self.text(message);
-                c::igEnd();
-                true
-            }
+    fn focus_current_window(&self) {
+        unsafe { c::igSetWindowFocus_Nil(); }
+    }
+
+    fn get_held_keys(&self, include_mouse: bool, include_modifiers: bool) -> Vec<imgui::Key> {
+        self.io()
+            .keys_down
+            .into_iter()
+            .skip(imgui::sys::ImGuiKey_NamedKey_BEGIN as _)
+            .enumerate()
+            .filter_map(|(index, pressed)| {
+                if pressed {
+                    if let Some(key) = index_to_imgui(index + imgui::sys::ImGuiKey_NamedKey_BEGIN as usize, include_mouse, include_modifiers) {
+                        Some(key)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+
+    fn begin_screen_cover(&self, focus: bool) -> imgui::WindowToken {
+        self.window("__cover")
+            .title_bar(false)
+            .resizable(false)
+            .movable(false)
+            .scroll_bar(false)
+            .collapsible(false)
+            .save_settings(false)
+            .focused(focus)
+            .size(self.io().display_size, imgui::Condition::Always)
+            .position([0.0, 0.0], imgui::Condition::Always)
+            .begin()
+            .expect("Error creating cover window")
+    }
+
+    fn popup_notification(&self, message: &str) -> bool {
+        let cover = self.begin_screen_cover(true);
+        if self.is_window_hovered() && self.is_mouse_clicked(imgui::MouseButton::Left) {
+            false
+        } else {
+            cover.end();
+            let screen_size = self.io().display_size;
+            self.window("Information")
+                .resizable(false)
+                .movable(false)
+                .scroll_bar(false)
+                .scrollable(false)
+                .collapsible(false)
+                .focused(true)
+                .position([screen_size[0] / 2.0, screen_size[1] / 2.0], imgui::Condition::Always)
+                .position_pivot([0.5, 0.5])
+                .build(|| {
+                    self.text(message);
+                });
+            true
         }
     }
+    
+    fn text_centered_float(&self, text: &str, center: Vec2<f32>) {
+        let size = self.calc_text_size(text);
+        let size = Vec2((size[0]+16.0) / 2.0, (size[1]+16.0) / 2.0);
+        let pos = [center.0 - size.0, center.1 - size.1];
+        self.window("__float")
+            .flags(imgui::WindowFlags::from_bits(0b100_0000_0011_1111_1111).unwrap()) // Note: previous code used the 0x20000 flag which doesn't seem to exist anymore. I'm not sure what that was for.
```

Comment:
Uhh. Aren't there any constants that could be used instead of this mystery value?
And could you please give a link to that "previous code"? I wasn't able to locate quickly what is meant here.

```suggestion
        self.window("__float") // Note: previous code used the 0x20000 flag which doesn't seem to exist anymore. I'm not sure what that was for.
            .flags(imgui::WindowFlags::from_bits(0b100_0000_0011_1111_1111).unwrap())
```